controller/engine_controller.go

```diff
@@ -1890,7 +1890,9 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replicaName, add
 					return
 				}
 
-				replica.Spec.FailedAt = util.Now()
+				now := util.Now()
+				replica.Spec.FailedAt = now
+				replica.Spec.LastFailedAt = now
```

Comment:
Let's start here.  With a correlated set of attributes, it's risky to set individual ones directly, allowing for possible contradictory situations if one is overlooked.  Better to have a "setter" method that does all together, perhaps in utils.  So setHealthyAt(time) would set not only Spec.HealthyAt and Spec.LastHealthyAt, but also clear FailedAt, if that's what we decide it should do.   Likewise for setFailedAt(time).

(With these changes in this improvement, non-empty FailedAt should mean "known to be failed right now" and should never be set at the same time that HealthyAt is, I claim. HealthyAt means "known to be healthy right now in a similar way.  It's a contradiction to have both non-empty, even if one could try to reason about the meaning of comparing the timestamps.  LastHealthAt and LastFailedAt are never cleared, so they make this crystal clear.

If I understand this right, it is not a contradiction for both HealthyAt and FailedAt to be empty.  That would be the case for a brand new replica (and LastHealthyAt and LastFailedAt would also be empty.  It would also be the case for a replica that is in the middle of rebuilding and has neither finished nor failed that process.  We could even put that test into a method isRebuilding(), and use it in place of the inline logic, which is used in (c *VolumeController) closeVolumeDependentResources() below, for instance.

Reply:
For an example of my confusion, see https://github.com/longhorn/longhorn-manager/blob/8a8414a150ca284996ac434160901f11e5f812fb/webhook/resources/replica/validator.go#L114

Reply:
I'm still working through this feedback a bit, but to start: I thought about the utility method to combine setting (for example) `HealthyAt` and `LastHealthyAt`. I ultimately decided against it because:

1. I don't think it really provides that much protection. Sure a developer could look and see if there is a utility function that they can use, but they may also just set whatever field they want independently.
2. (This is the primary reason I didn't.) We SHOULDN'T actually always set them together if we want to maintain the meaning of `HealthyAt`. I originally recognized that, but it looks like a bug snuck into this PR by which I accidentally set them both together in https://github.com/longhorn/longhorn-manager/blob/792a7e9f09c79760bc86409a66395b895dbcaa2a/controller/volume_controller.go#L721-L730. This was unintentional and incorrect. I intend to set `LastHealthyAt` EVERY time we see a RW replica and `HealthyAt` ONLY when we complete a rebuild.

Reply:
I did end up adding a utility function for `FailedAt` and `LastFailedAt` since these occur often and are atomic when setting to a time value.

Reply:
So, should the names be "LastRWAt" and "RebuiltAt"?